lexibank_satterthwaitetb.py

```diff
@@ -1,192 +1,261 @@
+            page_cells = page_cells[: -DROP_FINAL.get(page_cells[0], 1)]
+
+            # Once more due the PDF correction, in some cases the contents
+            # are shifted one entry to left due to empty cells. As the number
+            # of such empty cells changes from page to page, we need to
+            # test the rows, by index, to make sure the data is there.
+            page_rows = {}
+            for row_name, indexes in ROW_IDX.items():
+                if not indexes[0]:  # concept, first one
+                    idx2 = page_cells.index(indexes[1])
+                    row = page_cells[:idx2]
+                elif not indexes[1]:  # Zaozou, last one
+                    idx1 = page_cells.index(indexes[0])
+                    row = page_cells[idx1:]
                 else:
-                    row = cells[cells.index(indexes[0]): cells.index(indexes[1])]
+                    idx1 = page_cells.index(indexes[0])
+                    idx2 = page_cells.index(indexes[1])
+                    row = page_cells[idx1:idx2]
 
-                # if it is not the concept row, exclude the first element of
-                # each row (it is the language name, we already have it)
+                # In all cases except the concept row, exclude the first
+                # element (it is the language name, we already have it)
                 if row_name is not "concepts":
-                    rows[row_name] = row[1:]
+                    page_rows[row_name] = row[1:]
                 else:
-                    rows[row_name] = row
+                    page_rows[row_name] = row
 
-            page_data.append(rows)
+            # Finally collect the page data
+            page_data.append(page_rows)
 
-        with self.cldf as ds:
-            ds.add_sources()
-            ds.add_languages(id_factory=lambda l: slug(l["Name"]))
-            ds.add_concepts(id_factory=lambda c: slug(c.label))
+        return page_data
 
-            for page in pb(page_data):
-                # for each concept...
-                for idx, english in enumerate(page["concepts"]):
-                    for lang in page.keys():
-                        if lang in ["concepts", "Hanzi"]:
+    def cmd_makecldf(self, args):
+        # read page data, solving all PDF conversion problems (forms still
+        # need to be cleaned)
+        page_data = self._read_page_data()
+
+        # write the bibliographic sources
+        args.writer.add_sources()
+
+        # add the languages from the language list
+        for language in self.languages:
+            args.writer.add_language(
+                ID=slug(language["Name"]),
+                Name=language["Name"],
+                Glottocode=language["Glottocode"],
+                Glottolog_Name=language["Glottolog_Name"],
+            )
+
+        # add the concepts from the concept list
+        # TODO: add `English` and other fields?
+        for concept in self.conceptlist.concepts.values():
```

Comment:
Concept-class ideally follows what the concepticon list gives. So you may need a CustomConcept